src/messaging/tests/TestReliableMessageProtocol.cpp

```diff
@@ -0,0 +1,271 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file implements unit tests for the ExchangeManager implementation.
+ */
+
+#include "TestMessagingLayer.h"
+
+#include <core/CHIPCore.h>
+#include <messaging/ReliableMessageContext.h>
+#include <messaging/ReliableMessageManager.h>
+#include <protocols/CHIPProtocols.h>
+#include <support/CodeUtils.h>
+#include <transport/raw/tests/NetworkTestHelpers.h>
+
+#include <nlbyteorder.h>
+#include <nlunit-test.h>
+
+#include <errno.h>
+
+namespace {
+
+using namespace chip;
+using namespace chip::Inet;
+using namespace chip::messaging;
+
+using TestContext = chip::Test::IOContext;
+
+TestContext sContext;
+
+ReliableMessageManager manager;
+
+void test_os_sleep_ms(uint64_t millisecs)
```

Comment:
Generally tests should never sleep.

I believe RMP uses timers - could we provide a timer implementation to the functions, so that tests have a 'fake timer' that can be controller on when to fire by the tests themselves?

Reply:
It is not timer which is causing problem. The event-loop in system layer is not running, and I'm triggering the timer manually.

The problem is that there is no way to mock `System::Timer::GetCurrentEpoch()` function. I tried to make it not using the function, but as a result, all member functions of the manager class and the context class need to take an extra arguments `now`, I think it is an overshoot.

Currently the sleep time is limited to several milliseconds, and the test cases are running consistently w/o any problem. before we introduce a way to mock the function I think currently it is fine.